core/fetcher.go

```diff
@@ -0,0 +1,85 @@
+package core
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/celestiaorg/celestia-core/types"
+	"github.com/celestiaorg/celestia-node/service/block"
+)
+
+const newBlockSubscriber = "NewBlock/Events"
+
+var newBlockEventQuery = types.QueryForEvent(types.EventNewBlock).String()
+
+type BlockFetcher struct {
+	client Client
+
+	newBlockCh chan *block.Raw
+}
+
+// NewBlockFetcher returns a new `BlockFetcher`.
+func NewBlockFetcher(client Client) *BlockFetcher {
+	return &BlockFetcher{
+		client: client,
+	}
+}
+
+// GetBlock queries Core for a `Block` at the given height.
+func (f *BlockFetcher) GetBlock(ctx context.Context, height *int64) (*block.Raw, error) {
+	raw, err := f.client.Block(ctx, height)
+	if err != nil {
+		return nil, err
+	}
+	return raw.Block, nil
+}
+
+// SubscribeNewBlockEvent subscribes to new block events from Core, returning
+// a new block event channel on success.
+func (f *BlockFetcher) SubscribeNewBlockEvent(ctx context.Context) (<-chan *block.Raw, error) {
+	// start the client if not started yet
+	if !f.client.IsRunning() {
+		return nil, fmt.Errorf("client not running")
+	}
+	eventChan, err := f.client.Subscribe(ctx, newBlockSubscriber, newBlockEventQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a wrapper channel for translating ResultEvent to "raw" block
+	if f.newBlockCh != nil {
+		return nil, fmt.Errorf("new block event channel exists")
+	}
+
+	f.newBlockCh = make(chan *block.Raw)
```

Comment:
this one I'm still unsure about -- is it absolutely necessary?

Reply:
> is it absolutely necessary?

No

Reply:
and It's hard to explain this ;) 

You don't want to block constantly on writes to `newBlockCh` and make the routine wait once someone reads from it. Instead, it can be unblocked earlier if there is some buffer to write to, so it will continue processing the loop.

Basically, this is just a simple performance-enhancing trick that does not have any downsides. You might hear of things like lock-free data structures and else, so a small buffer here will remove unneeded locking.

Reply:
Can we leave this nit and implement it later if we feel like there is a need?

Reply:
Resolving this as we will think about it later whether we will need this optimisation.

Reply:
You can trust me that this works or leave it for later